### ISSUE TITLE
[Bug] PEP 484- and 585-compliant nested generics are broken af, bro

### ISSUE THREAD

**Issue by langfield** [beartype/beartype#140]
I think this will be easiest to understand from an example:
```python
from beartype import beartype

RowIndex = int 
ColIndex = int 

@beartype
def fn(a: RowIndex) -> ColIndex:
    return a

fn(RowIndex(1))
```
I want this call to `fn()` to fail (due to beartype, obviously). Clearly it can't be done in exactly the way I've written, so either this bit:
```python
RowIndex = int 
ColIndex = int 
```
or this bit:
```python
RowIndex(1)
```
must be done differently. I am open to any ideas! Of course we could write a class that subtypes `int`, but that is extremely clumsy and annoying. I want to do the aliasing and the instantiation from the supertype (i.e. the `RowIndex(1)`) in the least obtrusive way possible.

You could also use dataclasses, and then beartype will check the constructor, but then you have to do `row_idx.val` every time you want the value, which is highly annoying.

**Comment by leycec** [comment]
**Fun ensues!** Okay. Just to clarify, when you wrote "I want this call to `fn()` to fail (due to `beartype`, obviously)" about this line:

```python
fn(RowIndex(1))
```

...did you perhaps mean something like:

```python
# I want @beartype to raise *NO* exception here.
fn(RowIndex(1))

# I want @beartype to raise an exception here.
fn(1)
```

That is to say, do you want to ~~force~~ *encourage* the wonderful users of your beautiful API to explicitly instantiate a `RowIndex` object as if `RowIndex` were its own real class (which you of course poured time, forethought, lifeblood, and unit tests into) rather than simply passing an `int`? If so, you have a few miserable options available to you.

## NewType: Ain't So New, Ain't a Type

[**`typing.NewType`**](https://peps.python.org/pep-0484/#newtype-helper-function), first introduced with PEP 484 and supported by @beartype since forever, *basically* does everything you want – except the thing you most want. Notably:

* Static type checkers (e.g., `mypy`, `pyright`) will correctly enforce the constraint that callers *must* explicitly instantiate a `RowIndex` object rather than simply pass an `int`.
* Runtime type-checkers (so, @beartype is what we're saying) will **incorrectly ignore that constraint.** Why? Because `typing.NewType` was intentionally implemented to *not* be checkable at runtime. To check `typing.NewType` type hints at runtime, a runtime type-checker would need to dynamically inspect the abstract syntax tree (AST) parsed from Python source code. And... your app stack just consumed more wall-clock time than a crypto mining rig burning through its latest Nvidia card on a hot summer's day.

You'd use `typing.NewType` like so:

```python
from beartype import beartype
from beartype.typing import NewType

RowIndex = NewType('RowIndex', int)
ColIndex = NewType('ColIndex', int)

@beartype
def fn(a: RowIndex) -> ColIndex:
    return ColIndex(a)

# Both static and runtime type-checkers will
# raise *NO* error here.
fn(RowIndex(1))

# Static type-checkers will raise an error here.
# Sadly, runtime type-checkers will raise *NO* error here.
fn(1)
```

## A Fistful of Leftovers

This means we are now in the uncanny typing valley. Either your API needs to do something like...

> ...use dataclasses, and then `beartype` will check the constructor, but then you have to do `row_idx.val` every time you want the value, which is highly annoying.

**Agreed.** I'd rather eat week-old mashed potatoes with used chop sticks repeatedly licked by all three of our cats. That only leaves us with the nuclear typing option of...

> ...a class that subtypes `int`, but that is extremely clumsy and annoying.

**Yup.** This is it. Having exhausted all other options, we now scoop out the dregs from the coffee can of innovation:

```python
from beartype import beartype
from beartype.typing import NewType

class RowIndex(int): pass
class ColIndex(int): pass

@beartype
def fn(a: RowIndex) -> ColIndex:
    return ColIndex(a)

# Both static and runtime type-checkers will
# raise *NO* error here.
fn(RowIndex(1))

# Both static and runtime type-checkers will
# raise an error here.
fn(1)
```

Basically – you want the subclass approach, bro. As in so many things, explicit is better than implicit. This is less cumbersome than you might think. After all, what's less cumbersome than a one-liner? *Nuthin'*!

Like the other two approaches, it's still a one-liner. Unlike the other two approaches, it's well-behaved under both runtime *and* static type-checkers. The dark implicit magic of `typing.NewType` breaks runtime. That just leaves us with the blinding explicit truth of `class SubType(SuperType): pass`.

The only significant disadvantage of the subclass approach, as I see it, is **efficiency.** There's *probably* a negligible space and time penalty associated with doing that sort of thing. It's likely to be negligible, however. Profiling *probably* won't reveal any demonstrable difference between doing `RowIndex = int` and doing `class RowIndex(int): pass`.

What say ye, Victorian-era anti-industrialist poet? Is everything above still a satanic mill *or* did we kinda answer your apropos query?

Oh – and thanks for the continual interest. I *love* fielding fascinating QA conundrums like this. Let the type hints flow like fermented juniper berries, I say! :wine_glass:

**Comment by langfield** [comment]
> **Fun ensues!** Okay. Just to clarify, when you write "I want this call to `fn()` to fail (due to `beartype`, obviously)" about this line:
> 
> ```python
> fn(RowIndex(1))
> ```
> 
> ...do you perhaps mean something like:
> 
> ```python
> # I want @beartype to raise *NO* exception here.
> fn(RowIndex(1))
> 
> # I want @beartype to raise an exception here.
> fn(1)
> ```
> 
> That is to say, do you want to ~force~ _encourage_ the wonderful users of your beautiful API to explicitly instantiate a `RowIndex` object as if `RowIndex` were its own real class (which you of course poured time, forethought, lifeblood, and unit tests into) rather than simply passing an `int`? If so, you have a few miserable options available.

Surely it should also raise an exception for `fn(1)`, but note that the return type is annotated as `ColIndex`, so if we pass it a `RowIndex`, it should still be angry, because it is just returning its argument.

**Comment by langfield** [comment]
> Basically – you want the subclass approach, bro. As in so many things, explicit is better than implicit. This is less cumbersome than you might think. After all, what's less cumbersome than a one-liner? _Nuthin'_!

Damn, you are right, I do want the subclass approach, bro! I should have actually tried it and realized it could be done in one line before complaining!

**Comment by langfield** [comment]
Sadly, there is this
```python
class Board(List[List[str]]): pass

@beartype
def get_board() -> Board:
    return Board([1])

get_board()
```
No doubt you are not supposed to subclass subscripted type hints. But it's too bad this does not work.

The other annoyance is that anyone using `black` to autoformat code will not be able to define these aliases/classes in one line.

**Comment by leycec** [comment]
**Ohgodnoes.** Actually, you *are* supposed to subclass subscripted type hints. You're absolutely in the right and fully compliant with PEP typing standards like 484 and 585 here. You're doing good things.

It's @beartype that's at fault here – raising a non-human-readable exception at call time when returning an invalid object violating a PEP **484**-style nested sublist generic:

```python3
>>> from beartype import beartype
>>> from typing import List
>>> class Board(List[List[str]]): pass  # <----- PEP 484-style list generic
>>> @beartype
... def get_board() -> Board:
...    return Board([1])
>>> get_board()
Traceback (most recent call last):
  File "/home/leycec/tmp/mopy.py", line 16, in <module>
    print(f'get_board(): {get_board()}')
  File "<@beartype(__main__.get_board) at 0x7f8d13f8ab00>", line 40, in get_board
  File "/home/leycec/py/beartype/beartype/_decor/_error/errormain.py", line 317, in raise_pep_call_exception
    raise _BeartypeCallHintPepRaiseDesynchronizationException(
beartype.roar._BeartypeCallHintPepRaiseDesynchronizationException: @beartyped
__main__.get_board() return [1] violates type hint <class '__main__.Board'>, but
utility function raise_pep_call_exception() suggests this object satisfies this hint.
Please report this desynchronization failure to the beartype issue tracker
(https://github.com/beartype/beartype/issues) with the accompanying exception
traceback and this type object's representation:
[1]
```

That's bad, but at least an exception (however unreadable) is being raised. It could be worse... and it is! @beartype isn't raising *any* exception at call time when returning an invalid object violating a PEP **585**-style nested sublist generic:

```python3
>>> from beartype import beartype
>>> List = list
>>> class Board(List[List[str]]): pass  # <----- PEP 585-style list generic
>>> @beartype
... def get_board() -> Board:
...    return Board([1])
>>> get_board()
[1]
```

That's *really* bad. This is why I sigh with drooping cheeks and a crinkled brow. :face_exhaling: 

Something is rotten inside the @beartype codebase with respect to generics. Our extensive test suite covers both PEP 484- and 585-style generics – but maybe nested sublists like `class Board(List[List[str]])` expose an edge case we never tested?

**I know nothing.** I know this *should* work and possibly once worked, but certainly no longer works. Frankly, this is critical breakage. I'm immediately elevating this to a high priority bug. I'm putting on my blue-collar shirt, rolling up my grease-covered sleeves, and delving deep into the unkempt and rickety mine shaft that is the @beartype codebase.

If I survive this deadly ordeal, I should have this issue sufficiently beaten into submission within a few days. I'll then publish a new `beartype 0.10.5` point release resolving this – hopefully by this Friday. It's time to slay some mangy dragons, folks.

**Comment by langfield** [comment]
> **Ohgodnoes.** Actually, you _are_ supposed to subclass subscripted type hints. You're absolutely in the right and fully compliant with typing standards like PEP 484 and 585 here. You're doing good things.

Wow, my record is currently `0-2` for claiming things are true 'no doubt' this week. I should really stop saying that, lol.

Pleased to hear this is something you intend to support! Yeah, I had imported `List` with `from beartype.typing import List` so it was using the 585-style hints under the hood. No exception raised, as you said.

**Comment by leycec** [comment]
**Resolved by 673daa5.** We agree to pretend this issue never happened, I trust:

```python
$ python3.10
>>> from beartype import beartype

# It's PEP 484-style nested generics or bust.
>>> from typing import List
>>> class Board(List[List[str]]): pass  # <----- PEP 484-style list generic
>>> @beartype
... def get_board() -> Board:
...    return Board([1])
>>> get_board()
beartype.roar.BeartypeCallHintReturnViolation: @beartyped __main__.get_board()
return [1] violates type hint <class '__main__.Board'>, as generic base
typing.List[typing.List[str]] <protocol "__main__.Board"> index 0 item int 1 not
instance of list.

# It's PEP 585-style nested generics or bust, too.
>>> List = list
>>> class Board(List[List[str]]): pass  # <----- PEP 585-style list generic
>>> @beartype
... def get_board() -> Board:
...    return Board([1])
>>> get_board()
beartype.roar.BeartypeCallHintReturnViolation: @beartyped __main__.get_board()
return [1] violates type hint <class '__main__.Board'>, as generic base list[list[str]]
<protocol "__main__.Board"> index 0 item int 1 not instance of list.
```

It's a bit awkward that @beartype is claiming that generics are protocols in exception messages. Let's ignore that for now in the mutual interest of preserving my hair.

Thanks yet again for all the *wonderful* interest in runtime typing, @langfield. I'll try to push out `beartype 0.11.0` with all these fixes (and more!) by Friday. Until then, may you have an endless summer of beat poetry and unbeatable Python.
